### PR TITLE
[LLVM][VPlan] Pick more optimal initial value for VPBlend.

### PR DESCRIPTION
**Pull request by paulwalker-arm**
By choosing an initial value whose mask is only used by the blend we can remove the need for the mask entirely.

When going through my todo list I noticed my VPlanPatternMatch and OR simplification work (https://reviews.llvm.org/D157411) has been landed on main so here I'm updating https://reviews.llvm.org/D157412, which is now reimagined as a VPlan transformation.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 1]
```diff
@@ -918,6 +918,16 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
     // value with the others blended into it.
 
     unsigned StartIndex = 0;
+    for (unsigned I = 0; I != Blend->getNumIncomingValues(); ++I) {
+      // If a value's mask is only used by the blend then is can be deadcoded.
+      // TODO: Fine the most expensive mask that can be deadcoded.
```
Comment (david-arm) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:923]: nit: I think this should be `// TODO: Find`

**Review by david-arm — APPROVED** [review]
LGTM! Seems like a sensible improvement.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 2]
```diff
@@ -918,6 +918,16 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
     // value with the others blended into it.
 
     unsigned StartIndex = 0;
+    for (unsigned I = 0; I != Blend->getNumIncomingValues(); ++I) {
+      // If a value's mask is only used by the blend then is can be deadcoded.
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:922]:
```suggestion
      // If a value's mask is used only by the blend then is can be deadcoded.
```

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 3]
```diff
@@ -918,6 +918,16 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
...
+      // TODO: Find the most expensive mask that can be deadcoded.
+      VPValue *Mask = Blend->getMask(I);
+      if (Mask->getNumUsers() == 1 && !match(Mask, m_False())) {
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:925]: Suggest to remove all masked-off operands when starting to deal with Blend, so the rest can focus only on the rest. Independent of this patch.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 4]
```diff
@@ -996,6 +1006,7 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
                          m_LogicalAnd(m_VPValue(X1), m_Not(m_VPValue(Y1))))) &&
       X == X1 && Y == Y1) {
     R.getVPSingleValue()->replaceAllUsesWith(X);
+    R.eraseFromParent();
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1009]: Is this related to this patch or independent?
There are multiple RAUW's here, w/o such erasure, is this one special?

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 5]
```diff
@@ -918,6 +918,16 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
     // value with the others blended into it.
 
     unsigned StartIndex = 0;
+    for (unsigned I = 0; I != Blend->getNumIncomingValues(); ++I) {
+      // If a value's mask is only used by the blend then is can be deadcoded.
+      // TODO: Find the most expensive mask that can be deadcoded.
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:923]: Perhaps another TODO is to find a mask that are used only by multiple blends, which can deadcoded if handled together, along with a testcase having two or more suitable original phi's.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 4 above)
Comment (paulwalker-arm) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1009]: It's required.  In my view VPlanTransforms is being lazy with its cleanup as proven here because this simplification is crucial to how predicates are built up for tail-folded loops that employ if-conversion.  Without this change you get none of the expected benefit of the new vblend transformation because the mask's use counts are incorrect.

I'd like to follow up to see if "fixing" the other transforms results in further optimisations, but this was the minimum to solve my initial need.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 3 above)
Comment (paulwalker-arm) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:925]: Noted.  I'll create a followup PR.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 2 above)
Comment (paulwalker-arm) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp]: Done, but this feels artificial given both are legitimate english.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 5 above)
Comment (paulwalker-arm) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp]: I've updated the comment.  I'll see about creating a test case separately as it's not specifically relevant to this PR.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 4 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1010]: Probably best to wait until https://github.com/llvm/llvm-project/pull/93998 lands (work-list for simplifyRecipes), eager removal may remove recipes that may have been added to worklist earlier.

Once it has been switched to a work-list, we should be able to simplify, clean-up dead recipes and re-simplify the instructions that have any of their users removed.

As the other patterns at the moment aren't guarded by a use-count check, I don't think there should be differences at the moment from cleaning up early.

Comment (paulwalker-arm) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1010]: Unless there's a very strong reason I'd rather not wait because tail folded loops are being artificially hampered (as shown via several performance degradations when SVE is enabled) and the unit-tests show only benefits to this PR as it stand today.

Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:1010]: I don't think this needs to hold up the patch, just want to avoid updating additional places preemptively.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 5 above)
Comment (ayalz) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp]: All phi's of a block are converted to blends with the same edge masks, so looking for a mask with a single blend using it restricts this patch to single phi'd blocks only.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 3 above)
Comment (ayalz) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:926]: Thanks. Can be a followup or preparatory PR.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 5 above)
Comment (paulwalker-arm) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp]: Thanks @ayalz, this was very useful.  I've added a multi-phi test to if-conversion-nest.ll to highlight the TODO.
